torch/fx/experimental/proxy_tensor.py

```diff
@@ -277,7 +278,10 @@ def get_proxy_slot(
     return res
 
 def snapshot_fake(val: Tensor) -> Optional[Tensor]:
-    return val.detach()
+    # val.detach() will also eventually call fast_detach(),
+    # but this saves us a full trip into __torch_dispatch__
+    # (snapshot_fake is called a lot)
```

Comment:
the fact that this gives a lot more speedup than letting `val.detach()` go straight through `FakeTensor.__torch_dispatch__` makes think it's worth looking more into whether the C++ to python dispatching logic can maybe be sped up

Reply:
I was curious what part of this PR offered the most speedups, so I did some benchmarking. The snapshot_fake by itself is responsible for the vast majority of the perf wins. The `register_fast_op_impl(torch.ops.aten.detach.default)(fast_detach)` doesn't seem to do much

Reply:
I saw the same locally. The reason I was kind of surprised is:

(1) today, `snapshot_fake()` calls `detach(FakeTensor)` today (which is clearly slow)

(2) if we think most of the slowness comes from the fact that `detach(FakeTensor)` goes through several layers of python decomps, then `register_fast_op_impl(torch.ops.aten.detach.default)(fast_detach)` should have sped up `snapshot_fake()` too, but it did not.

(3) what is the difference between `return val.detach()` and `return `fast_detach(val)` in terms of perf (given that `fast_detach` is already registered to `FakeTensorMode`)? I think the main difference is that `val.detach()` also needs to (a) run through the dispatcher to eventually call `FakeTensorMode.__torch_dispatch__`, and (b) `__torch_dispatch__` eventually calls into `fast_detach()`.

So that makes me think that one of (a) or (b) is slow (and looking into the cpp-to-python dispatching logic might have some low hanging perf fruit)